speechbrain/lobes/models/discrete/SQ-Codec.py

```diff
@@ -0,0 +1,1497 @@
+        arr : numpy.ndarray
+            A 3D array of shape (B, N, D).
+        offset_size : int or None, optional
+            The offset size to be applied to each codebook slice (default is None).
+
+        Returns
+        -------
+        numpy.ndarray
+            A 1D array representing the flattened codebooks.
+        """
+        assert (
+            len(arr.shape) == 3
+        ), "Input array must have 3 dimensions [B, N, D]"
+        N, B, D = arr.shape
+        arr = arr.copy()
+        if offset_size is not None:
+            for n in range(N):
+                arr[n, :, :] += offset_size * n
+        flattened_arr = np.concatenate(arr, axis=1)
+        return flattened_arr
+
+    def encode(self, inputs):
+        """
+        Encodes the input audio tensor using the scalar codec and quantizes the output.
+
+        Parameters
+        ----------
+        inputs : torch.Tensor
+            Input audio tensor of shape (B, T) or (B, 1, T), where B is the batch size
+            and T is the length of the audio sequence.
+
+        Returns
+        -------
+        tuple
+            A tuple containing:
+            - torch.Tensor: The flattened and quantized encoded representation of the input.
+            - torch.Tensor: Quantized embedding.
+        """
+        if inputs.dim() == 2:
+            inputs = inputs.unsqueeze(1)
+        compressed = self.scalar_codec.encode(inputs)
+        chunks = compressed.chunk(self.n_codebook, dim=1)
+        codec_ls = []
+        for i, chunk in enumerate(chunks):
+            chunk = chunk.detach().cpu().numpy().astype(np.int32) + 1
+            tmp_codec = ternary_matrix_to_decimal(chunk)
+            codec_ls.append(tmp_codec)
+        codec_ls = np.array(codec_ls)
+        flat_codec = self._flatten_codebooks(codec_ls, self.dim_codebook)
+        flat_codec = torch.from_numpy(flat_codec).to(torch.int32)
+        return flat_codec, compressed
+
+    def decode(self, codes):
+        """
+        Decodes the quantized codes back into an audio tensor.
+
+        Parameters
+        ----------
+        codes : torch.Tensor
+            Quantized codes with shape (B, T).
+
+        Returns
+        -------
+        torch.Tensor
+            Reconstructed audio signal.
+        """
+        assert codes.dim() == 2
+        B, T = codes.shape
+        assert (
+            T % self.n_codebook == 0
+        ), "Length T must be divisible by n_codebook"
+        codes = codes.view(B, self.n_codebook, -1).permute(1, 0, 2)
+        for i in range(self.n_codebook):
+            codes[i, :, :] -= i * self.dim_codebook
+        emb_quant = []
+        for i in range(self.n_codebook):
+            tmp_list = decimal_to_ternary_matrix(codes[i, :, :], D=9) - 1
+            emb_quant.append(tmp_list)
+        emb_quant = torch.cat(emb_quant, dim=1)
+        out = self.scalar_codec.decode(emb_quant.float().to(self.device))
+        return out.detach().cpu().squeeze(0)
+
+    def infer(self, wav_root):
```

Comment:
Infer is not very self descriptive, especially since there is also an encode function. Maybe rename to same clearer?

Reply:
changed to "reconstruct"